[mono] Second try at using C11 atomics

lambdageek opened the pull request:
* For each platform decide if we will use C11 standard atomics, Win32 API atomics, GCC atomic intrinsics or emulated atomics.

   To use C11 atomics we generally want them to be lock-free for the primitive types we care about (that is, `ATOMIC_LONG_LONG_LOCK_FREE == 2` not `1` and similarly for other macros) because otherwise we cannot be sure if the atomic might use a global lock in which case we could have thread suspend problems if both the GC and the rest of the runtime use an atomic at the same time.

* On win32, use the win32 atomics until MSVC atomics support is not experimental, or we update our build to pass `/experimental:c11atomics`; and also until we build our C++ code with C++23 or later (otherwise MSVC will complain about including stdatomic.h)

* If the header gets included while we're generating offsets using `offsets-tool.py`, pretend we're using emulated atomics.  On some Linux configurations, the libclang that we use ends up picking up the platform `atomic.h` header, not the clang one, and then errors out on their underlying implementation.

* Replace the use of `bool` in some macros in Mono - it will expand to `_Bool` in C11 and mess things up.

   Use `boolean` instead

   Fixes https://github.com/dotnet/runtime/issues/91779

lambdageek committed:
[mono] Dont' use bareword `bool` in icall sig macros

It will expand to `_Bool` in C11 and mess things up.

Use `boolean` instead

Fixes https://github.com/dotnet/runtime/issues/91779

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek commented:
"Build linux-arm64 Release AllSubsets_Mono_LLVMFullAot_RuntimeTests llvmfullaot failed" is failing with

```
Running 'python3 /__w/1/s/src/mono/mono/tools/offsets-tool/offsets-tool.py --abi=aarch64-linux-gnu --netcore --targetdir="/__w/1/s/artifacts/obj/mono/linux.arm64.Release" --monodir="/__w/1/s/src/mono" --nativedir="/__w/1/s/src/native" --outfile="/__w/1/s/artifacts/obj/mono/linux.arm64.Release/cross/offsets-aarch-linux-gnu.h" --libclang="/usr/local/lib/libclang.so.16" --prefix="/crossrootfs/arm64/usr/lib/gcc/aarch64-linux-gnu/5" --sysroot="/crossrootfs/arm64"'
  Running clang: --target=aarch64-linux-gnu --sysroot /crossrootfs/arm64 -I /crossrootfs/arm64/include -I /crossrootfs/arm64/usr/lib/gcc/aarch64-linux-gnu/5/include -I /crossrootfs/arm64/usr/lib/gcc/aarch64-linux-gnu/5/include-fixed -std=gnu11 -DMONO_GENERATING_OFFSETS -I /__w/1/s/src/mono -I /__w/1/s/src/mono/mono -I /__w/1/s/src/mono/mono/eglib -I /__w/1/s/src/native -I /__w/1/s/src/native/public -I /__w/1/s/artifacts/obj/mono/linux.arm64.Release -I /__w/1/s/artifacts/obj/mono/linux.arm64.Release/mono/eglib -DTARGET_ARM64 -DHOST_LINUX -DMONO_CROSS_COMPILE -DUSE_MONO_CTX -DHAVE_SGEN_GC -DHAVE_MOVING_COLLECTOR /__w/1/s/src/mono/mono/metadata/metadata-cross-helpers.c
  
  /__w/1/s/src/mono/mono/utils/atomic.h:79:8: error: address argument to atomic operation must be a pointer to a trivially-copyable type ('volatile atomic_int *' (aka 'volatile _Atomic(int) *') invalid)
```

which looks similar to https://bugs.launchpad.net/ubuntu/+source/llvm-toolchain-6.0/+bug/1780747 which says that clang 6.0.1 had some kind of a regression.  Are we running some really old clang?

lewing commented:
should we revert the atomics pr while we sort this out?

lewing commented:
reverting in https://github.com/dotnet/runtime/pull/91812 while we figure out the failure

lambdageek commented:
I kind of suspect the LLVMFullAot lane is failing because it's using the GCC `stdatomic.h` header, not the clang `stdatomic.h` header.

In GCC `atomic_compare_exchange_strong` is a macro defined in terms of `__atomic_compare_exchange`.  And clang doesn't like `atomic_int *` as an argument type to that intrinsic:

Gnu header:
https://github.com/gcc-mirror/gcc/blob/b96b554592c5cbb6a2c1797ffcb5706fd295f4fd/gcc/ginclude/stdatomic.h#L168-L177

Clang output: https://godbolt.org/z/8WP7hbvG1

lambdageek committed:
Revert "Revert "[c11] define mono atomics in terms of standard atomics (#91489)" (#91812)"

This reverts commit 2157e377092116da30f1c8f2e9d41f23cea81166.

lambdageek committed:
Try to convince offset-tool to use clang's headers

lambdageek commented:
I merged `main` and reverted Larry's revert.  And also added `self.target_args += ["-isystem", clang_path + "/../lib/clang/16/include"]` to offsets-tool on linux-arm64 to try and convince it to get the clang header

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek commented:
@directhex do you think it's safe to do `clang_path = os.path.dirname(args.libclang)` `self.target_args += ["-isystem", clang_path + "/../lib/clang/16/include"]` on all platforms? I alaready do it for wasm and now for linux-arm64.

directhex commented:
@lambdageek it seems fair to try for _our_ infra. I can't guarantee it working on anyone else's.

lambdageek commented:
The include path didn't help on linux-arm64 llvmfullaot. Its clear what the problem is, but I have no idea how to fix the include paths without a repro environment.

lambdageek commented:
New approach.  The offsets tool doesn't actually need to look at anything in atomic.h, so just ifdef it out if we're generating offsets.

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek commented:
> New approach.  The offsets tool doesn't actually need to look at anything in atomic.h, so just ifdef it out if we're generating offsets.

Nope, no good. Other inline functions call these inline functions

am11 commented:
@lambdageek, does [this](https://discourse.llvm.org/t/compiling-with-built-in-stdatomic-vs-stdatomic-h/6225) help?

```c
// for C
#include <stddef.h>
#include <stdint.h>
#define OLD___STDC_HOSTED__ __STDC_HOSTED__
#undef __STDC_HOSTED__
#define __STDC_HOSTED__ 0
#include <stdatomic.h>
#undef __STDC_HOSTED__
#define __STDC_HOSTED__ OLD___STDC_HOSTED__

// for C++
// Pass `-stdlib=libc++` from py
```

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek committed:
new idea: if generating offsets, pretend we're using emulated atomics

the offset tool doesn't need to see the inlined atomic operations at all

Also, refactor the ifdef logic up front and define one of
MONO_USE_C11_ATOMIC, MONO_USE_WIN32_ATOMIC, MONO_USE_GCC_ATOMIC or
MONO_USE_EMULATED_ATOMIC depending on what we will use

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek commented:
I'm back to shaving this yak.  New approach: don't try to make the offsets tool smarter, just pretend we're going to use emulated atomics.  The offsets tool doesn't actually need to see the implementations of mono's atomic operations, so just avoid the whole problem.

lambdageek committed:
fixup - non-arm64 android

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek committed:
Revert "Try to convince offset-tool to use clang's headers"

This reverts commit edf18506932495dfff8a89ef4880ac3585b918e1.

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek commented:
@vargaz @jkoritzinsky @akoeplinger I assume your reviews are still 👍 ? Nothing substantially different - just some refactoring to not show offsets-tools the atomic implementations (it doesn't need to see function bodies)

lambdageek commented:
I believe the remaining failures are unrelated
